pkg/chain/ethereum/ethutil/ethutil.go

```diff
@@ -136,6 +136,37 @@ func CallAtBlock(
 	return nil
 }
 
+// EstimateGas tries to estimate the gas needed to execute a specific transaction based on
+// the current pending state of the backend blockchain. There is no guarantee that this is
+// the true gas limit requirement as other transactions may be added or removed by miners,
+// but it should provide a basis for setting a reasonable default.
+func EstimateGas(
+	from common.Address, 
+	to common.Address,
+	method string,
+	contractABI *abi.ABI,
+	transactor bind.ContractTransactor,
+	parameters ...interface{},
+) (uint64, error) {
+	input, err := contractABI.Pack(method, parameters...)
+	if err != nil {
+		return 0, err
+	}
+
+	msg := ethereum.CallMsg{
+		From: from,
+		To:   &to,
+		Data: input,
```

Comment:
We are not passing `Value` to `CallMsg` so the estimation may not be correct for `payable` functions. I think we need to generate a variation of estimate function just for payable functions.

It's not a top priority because the problem we are having in https://github.com/keep-network/keep-core/issues/1479 is solved with the code here. I'll open a separate issue to address payable functions.

Reply:
https://github.com/keep-network/keep-common/issues/26